### PR TITLE
Fixes #25548 - SearchBar bug on hosts page

### PR DESCRIPTION
**Pull request by Ron-Lavi**
**Description of problem:**

Unable to search host in search box for all host

**Steps to Reproduce:**
1. Go on all hosts page
2. click on search box and type to search

**Actual results:**
As soon as we try to search, search box is gets disappear with error on browser console.

**Expected results:**
User should able to search from available hosts.

### CODE REVIEW

**Comment by ares** [comment]
sorry, doesn't fix the problem on my setup :-(

**Comment by Ron-Lavi** [comment]
@ares Thanks, I didn't have the plugins environment yet,
so I just tried to fix it on blind.
on `Forklift centos7-devel` machine with katello there was no error,
trying to set up `centos7-luna-devel`, as I heard from coworkers that it fails there,
and than try to debug it and fix the problem.

**Comment by Ron-Lavi** [comment]
Finally ! @ares 
I succeeded to reproduce the bug on `vagrant@centos7-luna-devel`.
The problem was that the results was passed as an `Immutable array`,
and now when it is passed as `mutable` array, the bug is fixed.

@sharvit, @amirfefer Could you also take a look ?

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteSelectors.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteSelectors.js:1]: +1 for adding selectors.

Can you please add tests?

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteSelectors.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteSelectors.js]: Where do we actually break the `immutability`?
Can we change the consumer of the `results` array so it won't break the `immutability`?
If a 3rd party component breaks the `immutability`? If so can we pass the `results.asMutable()` only to the problematic component?

**Review by sharvit — CHANGES_REQUESTED** [review]
Nice work @Laviro :+1:

**Review by boaz0 — CHANGES_REQUESTED** [review]
No! Please don't do that.

You can read from the error message:
```
The push method cannot be invoked on an Immutable data structure
```

It seems that somewhere `push` is invoked on `results`! That's bad! The object is supposed to be immutable while `push` mutates the object.

I am against this :cry:

**Review by ohadlevy — DISMISSED** [review]
:+1: to what others said, this seems like we didn't find the root cause to the issue.

**Comment by sharvit** [comment]
@boaz1337 It seems that a 3rd party package do that and it is not apart of our source code.

If the 3rd party component changes the `results` array for its own inner usage and not in order to affect the state. I am fine with passing `asMutable` only to the relevant 3rd party property.

**Comment by sharvit** [comment]
Code wise looks good to me! Thanks, @Laviro :+1: 

@ares can you verify please?

File: webpack/assets/javascripts/react_app/components/AutoComplete/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/AutoComplete/index.js]: how about 
```js 
Immutable.isImmutable(results) ? results.asMutable() : results;
```
looks more clear to me

**Comment by boaz0** [comment]
@sharvit what library/package and where?

Returning a mutable object might be OK but a comment explaining that is a must. :newspaper_roll: 

Thanks. :dango:

**Comment by sharvit** [comment]
@boaz1337 I think it is the `TypeAheadSelect` from `patternfly-react`, @Laviro can you confirm?

I agree about the comment :+1:

**Comment by boaz0** [comment]
@Laviro Sweet heaven is it because of [react-bootstrap-typeahead](https://github.com/ericgio/react-bootstrap-typeahead)? :cloud_with_lightning:

**Comment by Ron-Lavi** [comment]
@boaz1337 yes, it happens while passing the Immutable array as results to the `react-bootstrap-typeahead`. While passing it as mutable everything works fine.
Not sure yet what causes it inside the `react-bootstrap-typeahead`.
Though this change really fix it.

**Comment by boaz0** [comment]
All mighty! Can we open a ticket? Or maybe open a pull request?
As much I want this fixed as soon as possible I have a feeling fixing it on `react-bootstrap-typeahead` side is smarter.

Anyway, please write a comment explaining why you're doing the `asMutable` magic (explaining that react-bootstrap-typeahead is mutating it before displaying it in the menu or something similar).

**Comment by ares** [comment]
Works great, fixes the issue for me. I think @boaz1337 last comment about adding a comment was not yet addressed, also linking upstream issue/PR that we've opened would be nice. Once I get ack from @boaz1337 I'm happy to merge, @ohadlevy won't be around so I'll have to dismiss his review :-)

**Comment by Ron-Lavi** [comment]
@amirfefer updated your request,
@boaz1337 added the comment, do you approve?
I'll open a PR to `react-bootstrap-typeahead` soon trying to fix this problem.

**Comment by boaz0** [comment]
Do you know in what version of `react-bootstrap-typeahead` this happens? I cannot reproduce it on `develop`.
Maybe it was fixed and all is needed is to update the package to have the newest version of that npm?

**Review by amirfefer — APPROVED** [review]
passing a mutable copied object from our store to a 3-rd party package seems fine to me.
Thanks @Laviro  works great :+1:

**Comment by Ron-Lavi** [comment]
@boaz1337 it happens on host page, you can reproduce it with `luna`,
I think it happens when passing a massive array with options to `react-bootstrap-typeahead`,
maybe than they try to manipulate the options.

anyway on our side there are no `push` method invoked on the array, 
and when passing the array as mutable to  `react-bootstrap-typeahead` it works perfect.

**Comment by Laviro** [comment]
[test upgrade]

**Comment by boaz0** [comment]
@Laviro - thanks for the details.
Right now, my resources are low so I am trying to reproduce it on `develop`.
You're saying if I have many hosts and I run a search query then the problem will appear, right?

**Comment by Ron-Lavi** [comment]
Yes @boaz1337, I double checked and pushed an immutable array of all countries as the default options,
and it happened again.

**Comment by Laviro** [comment]
[test upgrade]

**Comment by boaz0** [comment]
It looks good to me, just one small thing: 

can you specify the version of react-bootstrap-typeahead?
or have a link to the issue on react-bootstrap-typeahead?

That's all.

**Comment by Ron-Lavi** [comment]
thanks @boaz1337, patternfly is using `"react-bootstrap-typeahead": "^3.1.3",`
just opened an [issue](https://github.com/ericgio/react-bootstrap-typeahead/issues/398) in `react-bootstrap-typeahead`

**Review by boaz0 — APPROVED** [review]
LGTM

**Review by sharvit — APPROVED** [review]
LGTM! Thanks @Laviro :+1:

**Review by ares — APPROVED** [review]
Thanks @Laviro and all reviewers. @ohadlevy I'm merging since I know you'll be offline for few days. I believe the cause has been found and this is the best we can do for now in order to fix the issue.

**Comment by tbrisker** [comment]
1.20 :cherries: - 01e416a68

**Pull request merged**
